presence_data: Fetch presence data from the server

shuklamaneesh23 opened the pull request:
This PR addresses issue [#31039](#31039), where the app was not fetching presence data for users who had been idle for an extended period when the "Users" tab in the **Settings > Users** panel was opened. This caused outdated or missing availability data in the "Last active" column. With this update, the app now ensures that presence data is fetched for all users when the tab is opened, allowing for more accurate and up-to-date information to be displayed.

#### Key Updates:

1. **New Data Structure:**
   - Introduced a new structure, `user_setting_presence_info`, to store presence information specifically when the "Users" tab is opened in the settings panel. This structure helps track and display presence data for users, even if they have been idle for a long time.

2. **New API Call for Presence Data:**
   - Added a function `fetch_presence_user_setting`, which makes an API call to fetch presence data with the `history_limit_days` parameter set to `9999` so that it fetches all the availability data. This ensures that even long-idle users' data is retrieved and displayed accurately.


Fixes #31039.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
@mateuszmandera is this PR on the right track?

mateuszmandera commented:
I'm not sure this is the right approach, if I'm assuming our intent in #31039 correctly. This approach just makes it so that the webapp fetches 10 years of data on load. That wasn't our intent imo. We've chosen `settings.PRESENCE_HISTORY_LIMIT_DAYS_FOR_WEB_APP = 365` on purpose, and may want to adjust it at some point, but not for this issue.

#31039 is for fetching all the data **when Users tab of SETTINGS / Users is open**. So the webapp should make the appropriate request with a large `history_limit_days` param when the `Users` sections get opened. Not unconditionally on client load (which will significantly increase the demand on the server).

shuklamaneesh23 commented:
@mateuszmandera, should I pass `history_limit_days` as a high number like `3650` (10 years) only in the params of `/json/users/me/presence` when the Users tab in the settings is opened? Would that solve the issue? If so, should I roll back the other changes I made in different files, or is there another direction I should consider?

mateuszmandera commented:
Yes, I think so - the app should just make a separate request like that, as outlined in the brief discussion in https://chat.zulip.org/#narrow/stream/9-issues/topic/last.20active.20time - the other changes don't seem necessary  for that

shuklamaneesh23 commented:
Hey @mateuszmandera ,I have undone all the unnecessary changes now, can you please review it again.

mateuszmandera commented:
This change doesn't accomplish the goal - you're just adding the param to presence requests that are made by the app at regular intervals. It has to make a separate request with this param when opening `Users`.

shuklamaneesh23 commented:
Hey @mateuszmandera ,could you please check my PR code changes and tell me whether I am right track for solving this issue as currently some tests are failing.
Thanks!

mateuszmandera commented:
Looks like a step in the right direction. I'm not super able to give a good review on the details of the frontend code that's not directly dealing with the presence API and will just defer to someone else on that if the core of this works well, but this now looks along like lines of what I had in mind. Have you tested manually whether this works and behaves as expected?

mateuszmandera commented on web/src/activity.ts:
```diff
@@ -8,6 +8,29 @@ import {page_params} from "./page_params";
+    {active_timestamp: number; idle_timestamp: number}
+> = {};
+
+// Define the schema for the response of the POST request to fetch the presence data of last 10 years.
+const post_presence_user_settings_response_scheme = z.object({
+    result: z.string(),
+    msg: z.string(),
+    presences: z
+        .record(
+            z.string(),
+            z.object({
+                active_timestamp: z.number(),
+                idle_timestamp: z.number(),
+            }),
+        )
+        .optional(),
+    server_timestamp: z.number().optional(),
+    presence_last_update_id: z.number().optional(),
+});
```
I don't think you have to duplicate this schema, you should probably just be able to re-use `post_presence_response_schema` here?

shuklamaneesh23 replied:
Thanks for pointing out this..I will remove this redundant part.

mateuszmandera commented on web/src/activity.ts:
```diff
@@ -166,6 +189,50 @@ export function send_presence_to_server(redraw?: () => void): void {
     });
 }
 
+export async function fetch_presence_user_setting(): Promise<void> {
```
Can you explain why this function and `populate_users()` in the next commit have to be made async?

shuklamaneesh23 replied:
Since I am fetching the presence data using this function, where I call the API endpoint, as you suggested in your previous comments, the API should only be called when the "Users" tab of the settings is used. If I do not make it asynchronous, the data is not displayed correctly on the frontend, as it calculates the `last_active_data` from an empty data structure because the API has not yet fetched the data. I want `populate_users` to wait until the data is fetched and then proceed with rendering the active users table on the frontend.

Therefore, I introduced `async` and used `await` in `populate_users` because `populate_users` is responsible for creating the `active_users_table`. When it is called, it will first fetch the data for the past 10 years or more, and only then display the table on the frontend.

shuklamaneesh23 commented:
> Looks like a step in the right direction. I'm not super able to give a good review on the details of the frontend code that's not directly dealing with the presence API and will just defer to someone else on that if the core of this works well, but this now looks along like lines of what I had in mind. Have you tested manually whether this works and behaves as expected?

Yes @mateuszmandera , I have tested it manually on my localhost and it is working as expected.

shuklamaneesh23 commented:
Hey @mateuszmandera , can you please review this now as I have made the above suggested changes and now all the tests are passing.
Thanks!

mateuszmandera commented on web/src/settings_users.js:
```diff
@@ -633,3 +643,44 @@ export function set_up_bots() {
+                status: compute_active_status(),
+                ping_only: false,
+                last_update_id: -1,
+                history_limit_days: 3650,
```
Rather than framing this as a "10 years of history", I think you should just pass an even greater value, like 9999 to fully express the intent that we just want all history (not some arbitrary cutoff).

shuklamaneesh23 replied:
sure!

mateuszmandera commented on web/src/settings_users.js:
```diff
@@ -633,3 +643,44 @@ export function set_up_bots() {
+                    const data = post_presence_response_schema.parse(response);
+
+                    if (data.presences) {
+                        const updatedPresenceInfo = {};
```
You're mixing camel case naming convention with our usual snake case. We want to be consistent, so this should be `updated_presence_info`.

shuklamaneesh23 replied:
ok, I will make sure not to do this case mistake in future.
Thanks!

mateuszmandera commented on web/src/settings_users.js:
```diff
@@ -633,3 +643,44 @@ export function set_up_bots() {
+                    if (data.presences) {
+                        const updatedPresenceInfo = {};
+
+                        for (const [userId, presenceData] of Object.entries(data.presences)) {
```
same here, `user_id` and `presence_data`.

mateuszmandera commented:
okay, I tested this a bit too, and it seems to work fine indeed. I posted some suggested tweaks above, but in general, as far as the general mechanism and use of the API is concerned (which is what I'm able to review), this looks fine to me.

@amanagr Can you perhaps review the code more generally? Like the use of `async` and other conventions; I don't feel I have a good grasp of this kind of web app code.

shuklamaneesh23 commented:
Thank you, @mateuszmandera , for providing such a detailed review. I have implemented the suggested changes accordingly.

amanagr commented on web/src/settings_users.js:
```diff
@@ -176,18 +180,24 @@ function create_role_filter_dropdown($events_container, section) {
-    create_role_filter_dropdown($("#admin-user-list"), section.active);
-    create_role_filter_dropdown($("#admin-deactivated-users-list"), section.deactivated);
+    try {
+        await fetch_presence_user_setting();
+        section.active.create_table(active_user_ids);
+        section.deactivated.create_table(deactivated_user_ids);
+        create_role_filter_dropdown($("#admin-user-list"), section.active);
+        create_role_filter_dropdown($("#admin-deactivated-users-list"), section.deactivated);
```
We don't need to use an async function here. Let's just pass these additional steps post fetch in the success callback.

amanagr commented on web/src/settings_users.js:
```diff
@@ -633,3 +643,44 @@ export function set_up_bots() {
+                }
+            },
+            error(error) {
+                throw new Error("Failed to fetch presence data: " + error);
```
Do we really want to throw an error here? I think we should just display users using the information we have instead.

shuklamaneesh23 replied:
Should I use redraw here or just remove this error throwing part from here?

shuklamaneesh23 commented:
Thanks @amanagr for those suggestions that how can I remove that async. I have implemented those and the PR is again up for review.
Thanks!

mateuszmandera commented on web/src/settings_users.js:
```diff
@@ -263,12 +263,16 @@ function bot_info(bot_user_id) {
-    if (!last_active_date) {
-        return $t({defaultMessage: "Unknown"});
+    if (user_setting_presence_info[String(user.user_id)]) {
+        const userPresence = user_setting_presence_info[String(user.user_id)];
```
As I noted earlier, you should stick to snake_cave. So `user_presence`.

shuklamaneesh23 replied:
Got it, I'll update the variable name to `presence_row` as suggested. Thanks for pointing it out!

mateuszmandera commented:
Hmm and it looks like the function is still `async`? Unless I'm misunderstanding what Aman was proposing

shuklamaneesh23 commented:
Hey @mateuszmandera, I've removed the `async` from the function where Aman pointed out and replaced it with a `success` callback as suggested. The rest of the function remains `async`.

amanagr commented:
We don't need to keep the function `async` for no reason, we don't have a pattern of using `async` functions in the codebase, so it is easier to reuse  and debug without it being `async`.

shuklamaneesh23 commented:
Ok @amanagr , should I use `promise` without using async await so that i can use that success call back at that place where you mentioned before.

shuklamaneesh23 commented:
> We don't need to keep the function `async` for no reason

The reason I kept `async` here is that, without it, the API call would render the list without waiting for the endpoint to fetch the user's presence details.

> We don't have a pattern of using `async` functions in the codebase

I have noticed some `async` functions in the codebase. For example, the file `/web/e2e-tests/lib/common.ts` uses a similar structure for writing `async` functions with promises.

shuklamaneesh23 commented:
Hey @amanagr , I have removed the async from the function definition. Could you please check it again and let me know what else I need to fix. 
Thanks!

amanagr commented:
We don't need to use `Promise` here too. Just move the `then` function methods to success callback too.

shuklamaneesh23 commented:
Hello @amanagr , I have updated the code according to the previous comment.
Thanks!

amanagr commented:
I think @sahil839 is the right person to review this.

sahil839 commented on web/src/settings_users.js:
```diff
@@ -25,6 +27,8 @@ import * as user_sort from "./user_sort";
 export const active_user_list_dropdown_widget_name = "active_user_list_select_user_role";
 export const deactivated_user_list_dropdown_widget_name = "deactivated_user_list_select_user_role";
+// introduced a new data structure to store the presence data of users for the last 10 years.
+export const user_setting_presence_info = {};
```
Can we directly use `Map` here like we do for `presence_info` in `presence.ts`.

sahil839 commented on web/src/settings_users.js:
```diff
@@ -256,12 +263,16 @@ function bot_info(bot_user_id) {
+        return timerender.render_now(last_active_date).time_str;
     }
-    return timerender.render_now(last_active_date).time_str;
+    return $t({defaultMessage: "Unknown"});
```
I guess would be better to have this block of returning "Unknown" inside `if` condition for readability.

sahil839 reviewed (commented):
Posted a couple of minor comments.

shuklamaneesh23 commented:
Thank you @sahil839 for the suggestions. The PR is now updated and ready for review again.

Thanks!

sahil839 commented on web/src/settings_users.js:
```diff
@@ -263,11 +263,17 @@ function bot_info(bot_user_id) {
 
 function get_last_active(user) {
-    const last_active_date = presence.last_active_date(user.user_id);
+    const presence_row = user_setting_presence_info.get(String(user.user_id));
```
I don't think the `String` conversion is needed.

shuklamaneesh23 replied:
@sahil839, I tried that, but it always results in "Unknown," which means the `presence_row` value isn’t being fetched correctly without using a string.

mateuszmandera commented on web/src/settings_users.js:
```diff
@@ -631,3 +638,37 @@ export function set_up_bots() {
+            if (data.presences) {
+                user_setting_presence_info.clear();
+
+                for (const [user_id, presence_data] of Object.entries(data.presences)) {
```
```suggestion
                for (const [user_id_str, presence_data] of Object.entries(data.presences)) {
                    const user_id = Number.parseInt(user_id_str, 10);
```

you can just do this like in https://github.com/zulip/zulip/blob/94a86d4ac2b68fd7efa3a3bc0b3ee4f0cb18e3f7/web/src/presence.ts#L197-L198 to be consistent with that style and address @sahil839's comment in https://github.com/zulip/zulip/pull/31895#discussion_r1816150382

sahil839 replied:
Ohhk, I missed that we parse it to integer in `presence.ts`.

shuklamaneesh23 commented:
Thanks, @mateuszmandera and @sahil839! I've updated the code based on the feedback you provided.

sahil839 commented:
Looks good to me.

shuklamaneesh23 commented:
Hi @sahil839, Apologies for the ping. I just wanted to check if the PR is ready for an integration review yet.
Thank you!

timabbott commented on web/src/settings_users.js:
```diff
@@ -631,3 +638,38 @@ export function set_up_bots() {
+            status: compute_active_status(),
+            ping_only: false,
+            last_update_id: -1,
+            history_limit_days: 9999,
```
I would write `365 * 1000` here so it's more obviously forever.

timabbott commented on web/src/settings_users.js:
```diff
@@ -631,3 +638,38 @@ export function set_up_bots() {
+                    user_setting_presence_info.set(user_id, {
+                        active_timestamp: presence_data.active_timestamp,
+                        idle_timestamp: presence_data.idle_timestamp,
+                    });
```
There's a few things that could be improved with this logic:
- We should only ever need to do this sort of full fetch once, not every time we load the "Users" panel. Likely some sort of state variable to track if we've fetched this would be appropriate.
- Ideally, we'd be able to use these data in other UI elements that display presence, not just here. I think we can just store these data via `presence.set_info(data.presences, data.server_timestamp, data.last_update_id)`, with a comment explaining that the next regular default presence checkin with the server should naturally pick up from here.
- I think it might be nice to have the "Users" table load immediately and just show loading indicators in the relevant column for any users for whom we don't have presence data, and trigger a rerender when this request returns. Is that an easy restructuring?

shuklamaneesh23 replied:
@timabbott, from your comment, I understand that I may not need to introduce an extra variable `user_setting_presence_info` here. Instead, I should fetch the complete data on the initial load of the application and then continue updating it through regular `update-user-presence` calls, using this data wherever we need to display presence information.

Am I understanding this correctly?

> Is that an easy restructuring?

I'm not entirely sure, but I’ll give it a try and see how it goes. 

Thank you!

shuklamaneesh23 replied:
> * We should only ever need to do this sort of full fetch once, not every time we load the "Users" panel. Likely some sort of state variable to track if we've fetched this would be appropriate.

Introduced a presence_data_fetched flag to keep track of whether this data has been fetched.

> * Ideally, we'd be able to use these data in other UI elements that display presence, not just here. I think we can just store these data via `presence.set_info(data.presences, data.server_timestamp, data.last_update_id)`, with a comment explaining that the next regular default presence checkin with the server should naturally pick up from here.

Implemented this approach and removed the extra variable I initially introduced.

> * I think it might be nice to have the "Users" table load immediately and just show loading indicators in the relevant column for any users for whom we don't have presence data, and trigger a rerender when this request returns. Is that an easy restructuring?

 I found this restructuring challenging. I attempted it but was not successful.

@timabbott, could you please review and let me know if I’ve correctly addressed the first two points?
Thanks!

shuklamaneesh23 commented:
Hello @timabbott , I have pushed some changes. Could you please review that?

mateuszmandera commented:
@shuklamaneesh23 It'll be helpful if you mention explicitly what changes of the suggested ones you made (unless it's all of them)? Right now it's not super clear if the `it might be nice to have the "Users" table load immediately and just show loading indicators in the relevant column` suggestion is implemented or skipped. If it is, then it'd be good to include a screenshot (or probably screen recording) demonstrating how the new behavior with loading looks to a user.

shuklamaneesh23 commented:
@mateuszmandera , I have mentioned in the comment itself that i haven't implemented that. 

Rest of the suggestions were implemented. 

Thanks!

shuklamaneesh23 commented:
Hello @mateuszmandera  and @timabbott ,

I have addressed all the suggestions.
Here, is the demo screen recording of what @mateuszmandera has asked for:

<details>
<summary>Video demonstrating new behaviour</summary>

https://github.com/user-attachments/assets/328b8bad-35ad-4aef-aa50-88bc344b6985


</details>

shuklamaneesh23 commented:
Hello @timabbott,

Could you please take a look and let me know if the fixes I made are as per your suggestions, or if further adjustments are needed?
Thanks!

timabbott commented on web/src/settings_users.ts:
```diff
@@ -554,12 +570,6 @@ export function redraw_active_users_list(): void {
-    });
-    loading.make_indicator($("#admin_page_deactivated_users_loading_indicator"), {
-        text: $t({defaultMessage: "Loading…"}),
-    });
```
Why do you remove all of this loading indicator code? 

It seems very unlikely to be correct. Likely you instead want to call the function to hide those indicators inside `render_table`, and probably also inside an `error` handler for the `channel.post` call.

shuklamaneesh23 replied:
I removed these `loading indicators` because you mentioned, `it might be nice to have the 'Users' table load immediately and just show loading indicators in the relevant columns` where availability data is not available initially, and then update those values via a call to the new API endpoint. Therefore, these loading indicators seemed unnecessary. 

Please correct me if I misunderstood your suggestion.

Please have a look at the new behaviour .
Thanks!
<details>
<summary>New Behaviour: </summary>

https://github.com/user-attachments/assets/3ddb36ce-ff34-4097-9fc1-8d15b0d4605c


</details>

Would it be a good idea to keep them as they were previously?

timabbott replied:
OK but you didn't implement the new location for loading indicators.

Look, if the user is waiting for the app to fetch data from somewhere, there should be a loading indicator, rather than it just silently not showing the data. This is basic UX design. I suggested a nice improvement might be to have the loading indicator, rather than being global, be just in the place where there's data missing. But under no circumstances does it make sense to just remove an existing working set of loading indicators.

timabbott replied:
The video does seem to have indicators in the new place, but they're just the word "Loading..." -- you should use the `loading.ts` spinners that look nice. I think that not being there was the source of my confusion.

shuklamaneesh23 replied:
@timabbott, I have restored the previous spinners. However, I haven't yet implemented the spinner alongside `Loading...` in each row. I was trying to figure out how to add that spinner, but it was taking time. I can either add a commit or create a follow-up PR—whichever you prefer—to include the spinner in each row once my college exams are over.

Thanks!

timabbott replied:
Yeah that can be a follow-up project.

timabbott commented on web/src/settings_users.ts:
```diff
@@ -228,8 +231,23 @@ function populate_users(): void {
     if (active_user_ids.length === 0 && deactivated_user_ids.length === 0) {
         failed_listing_users();
+        return;
     }
```
@sahil839 is this block a case that can actually happen? I'd expect you can always see yourself here.

sahil839 replied:
Yes, there would be atleast one active user, i.e. yourself. So, we can remove this.

shuklamaneesh23 replied:
@sahil839 , should I remove this by making a follow up PR?

timabbott commented:
Merged, thanks @shuklamaneesh23 and @sahil839!
